feat(macro): improve argument parsing

ylemkimon opened the pull request:
**1. Primitive v.s. Macro**
LaTeX has two types of command (control sequence): TeX primitive and macro. TeX primitives are provided by TeX engine. Macros, in contrast, are defined using primitives in the LaTeX kernel (`latex.ltx`) and packages such as `amsmath`. (Macro may expand into primitive commands without taking arguments. They are considered as primitives hereinafter.) Primitives scan argument as it goes, i.e., expanding macros, whereas macros take a properly nested group or non-blank token without expanding. For instance, 
```tex
\def\foo{12}
\sqrt\foo
\underline\foo
```
is equivalent to
```tex
\sqrt12
\underline{12}
```

However, KaTeX uses primitive argument parsing for functions. This PR introduces a new argument type `primitive`, which uses the existing method, and uses it for parsing arguments of primitive commands. It also changes other group parsers such as `parseArgumentGroup` (new) and `parseStringGroup` to use macro argument parsing method, which scans the argument and denotes the end of the argument at the `MacroExpander` level before parsing. This fixes #924, fixes #1027, fixes #1720, and fixes #2319. Closes #611.

**2. Macro**
Delimited parameters in macro and brace handling are implemented in the new `consumeArg` function. Braced control sequence and unbraced replacement text no longer works.[1]

Optional arguments are implemented using a delimited parameter at the `MacroExpander` level. This fixes #1419 and #2288.

Whitespace handling is also improved to consume only in undelimited macro arguments and required/optional arguments. This broke `\\` and this PR fixes it along with fix #2119.

---

Some changes may result in **BREAKING CHANGEs** such as [1], and issues linked above, however they probably were originally wrong behavior.

ylemkimon commented on src/MacroExpander.js:
```diff
@@ -108,6 +109,30 @@ export default class MacroExpander implements MacroContextInterface {
+
+        // indicate the end of an argument
+        const pos = this.lexer.tokenRegex.lastIndex;
+        this.pushToken(new Token("EOF", new SourceLocation(this.lexer, pos, pos)));
```
TODO: find a better way to indicate the end of an argument

ylemkimon commented on src/MacroExpander.js:
```diff
@@ -122,42 +147,82 @@ export default class MacroExpander implements MacroContextInterface {
+                    ", expected '" + (delims && isDelimited ? delims[match] : "}") +
+                    "'", tok);
+            }
+            if (delims && isDelimited) {
```
`delims && ` is required for Flow to recognize `delims != null`.

ylemkimon commented on src/Parser.js:
```diff
@@ -476,31 +475,34 @@ export default class Parser {
-                if (isOptional) {
+
+            // \sqrt is a macro only if an optional argument exists
+            if (funcData.type === "sqrt" && i === 1 && optArgs[0] == null) {
```
I can't seem to find a better way to handle `\sqrt` than hard-coding it.

ylemkimon commented on src/functions/kern.js:
```diff
@@ -14,7 +14,7 @@ defineFunction({
     props: {
         numArgs: 1,
-        argTypes: ["size"],
+        argTypes: ["primitive_size"],
```
BREAKING CHANGE: braced size arguments no longer work.

ylemkimon commented on src/macros.js:
```diff
@@ -198,39 +204,59 @@ defineMacro("\\char", function(context) {
-    let arg = context.consumeArgs(1)[0];
-    if (arg.length !== 1) {
-        throw new ParseError("\\gdef's first argument must be a macro name");
+    let tok = context.popToken();
```
BREAKING CHANGE: braced control sequence no longer works.

ylemkimon commented on src/macros.js:
```diff
@@ -198,39 +204,59 @@ defineMacro("\\char", function(context) {
     }
-    // Final arg is the expansion of the macro
+    // replacement text, enclosed in '{' and '}' and properly nested
+    const tokens = context.consumeArg(null, preserveOutermostBraces);
```
BREAKING CHANGE: unbraced replacement text no longer works

ylemkimon commented on src/Parser.js:
```diff
@@ -759,6 +704,36 @@ export default class Parser {
+        const result = {
+            type: "ordgroup",
+            mode: this.mode,
+            loc: SourceLocation.range(firstToken, lastToken),
```
This may be a regression, as `firstToken` is the first token after the opening brace, if it exists.

ylemkimon commented on src/functions/genfrac.js:
```diff
@@ -444,7 +444,7 @@ defineFunction({
     props: {
         numArgs: 1,
-        argTypes: ["size"],
+        argTypes: ["primitive_size"],
```
BREAKING CHANGE: braced size arguments no longer work.

ylemkimon commented on src/Parser.js:
```diff
@@ -698,19 +651,22 @@ export default class Parser {
+        if (res.text.length === 0) {
             // Because we've tested for what is !optional, this block won't
             // affect \kern, \hspace, etc. It will capture the mandatory arguments
             // to \genfrac and \above.
```
*

ylemkimon commented:
This breaks the example at the main page:
```tex
% \f is defined as f(#1) using the macro
\f{x} = \int_{-\infty}^\infty
    \hat \f\xi\,e^{2 \pi i \xi x}
    \,d\xi
```
Any suggestion to replace it is welcome!

ylemkimon commented on src/macros.js:
```diff
@@ -198,39 +214,63 @@ defineMacro("\\char", function(context) {
-        throw new ParseError("\\gdef's first argument must be a macro name");
+    let tok = context.popToken();
+    const name = tok.text;
+    if (name[0] !== "\\") {
```
TODO: allow active character macro

kevinbarabash commented:
> This breaks the example at the main page

Is there any way to change the example to make it work?  I tried the following:
```
\def\f#1{f(#1)}
\f{x} = \int_{-\infty}^\infty
    \hat \f\xi\,e^{2 \pi i \xi x}
    \,d\xi
```
on the deploy preview and got:
```
KaTeX parse error: Unexpected end of input in a macro argument, expected '}' at position 97: …infty \hat \̲f̲\xi\,e^{2 \pi i…
```
Quicklatex.com also produces an error which is a good sign.  My TeX-fu isn't very strong unfortunately.  `\def\f#1{f(#1)}` seems reasonable, but apparently it is not. :(

kevinbarabash commented:
I figured out the issue.  We have to use explicit groups for both `\f` and `\hat`.  The following works:
```
\def\f#1{f(#1)}
\f{x} = \int_{-\infty}^\infty
    \hat{\f\{xi\}},e^{2 \pi i \xi x}
    \,d\xi
```

kevinbarabash commented:
I'm glad to see that we closing more gaps between KaTeX and LaTeX.  I think this change is going to require some docs on how to migrate existing TeX that people may have written that will no longer work after these changes.

kevinbarabash commented:
What I typed before was wrong, I meant to write:
```
\def\f#1{f(#1)}
\f{x} = \int_{-\infty}^\infty
    \hat{\f{xi}},e^{2 \pi i \xi x}
    \,d\xi
```
Unfortunately, this doesn't give the output we want:
<img width="826" alt="Screen Shot 2019-12-31 at 6 10 11 PM" src="https://user-images.githubusercontent.com/1044413/71636031-d09c2800-2bf8-11ea-9832-d7650a8cc705.png">

On the bright side this is also what LaTeX does:
<img width="672" alt="Screen Shot 2019-12-31 at 6 11 10 PM" src="https://user-images.githubusercontent.com/1044413/71636045-0c36f200-2bf9-11ea-919c-cc31f2aef1c8.png">

What if we change `\def\f#1{f(#1)}` to `\def\func#1#2{#1(#2)}`?  Then the example would look like:
```
% \func is a macro defined to be {#1(#2)}
\func{f}{x} = \int_{-\infty}^\infty
    \func{\hat f}{xi}\,e^{2 \pi i \xi x}
    \,d\xi
```
The result has the hat in the right spot.
<img width="833" alt="Screen Shot 2019-12-31 at 6 16 34 PM" src="https://user-images.githubusercontent.com/1044413/71636100-c62e5e00-2bf9-11ea-9858-f4e170ef751d.png">

ylemkimon commented:
I've updated `\f` to take two arguments to demonstrate a macro can take multiple arguments and a control sequence as an argument. I wanted to demonstrate KaTeX supports a delimited macro, but it's currently not possible to define a delimited macro. But I think this is out of this PR's scope. I've also added a migration guide.

Whadup commented:
I'm really looking forward to this merge =)

edemaine commented:
We should undo #2289 in this PR. 🙂

kevinbarabash commented:
@ylemkimon in #2288 you commented:
> Actually, #2085 will add support for delimited macros.

but earlier in this PR you also commented:
> I wanted to demonstrate KaTeX supports a delimited macro, but it's currently not possible to define a delimited macro. But I think this is out of this PR's scope. I've also added a migration guide.

Looking at the code it looks delimiters are supported.  Is that correct?  If so, could you add some tests that demonstrate this?

ylemkimon commented:
@kevinbarabash I meant it's not possible to define a delimited macro using `options.macros`.

ylemkimon commented on test/katex-spec.js:
```diff
@@ -3173,32 +3166,36 @@ describe("A macro expander", function() {
-        //expect`\gdef{\foo}{}`.not.toParse();
+    });
+
+    it("\\gdef defines macros with delimited parameter", function() {
+        expect`\gdef\foo|#1||{#1}\text{\foo| x y ||}`.toParseLike`\text{ x y }`;
+        expect`\gdef\foo#1|#2{#1+#2}\foo 1 2 |34`.toParseLike`12+34`;
+        expect`\gdef\foo#1#{#1}\foo1^{23}`.toParseLike`1^{23}`;
+        expect`\gdef\foo|{}\foo`.not.toParse();
+        expect`\gdef\foo#1|{#1}\foo1`.not.toParse();
+        expect`\gdef\foo#1|{#1}\foo1}|`.not.toParse();
```
Tests for delimited macro are added here.

kevinbarabash replied:
I was wondering what the `|` were for.  I'm so used to brackets and parens as delimiters, I forgot that `|` is a delimiter as well. 🙃

kevinbarabash commented on src/MacroExpander.js:
```diff
@@ -122,42 +148,89 @@ export default class MacroExpander implements MacroContextInterface {
+            tokens.shift();
+        }
+        tokens.reverse(); // to fit in with stack order
+        return {tokens, start, end: tok};
```
`start` and `end` can be outside the range of `tokens`, correct?

ylemkimon replied:
@kevinbarabash No, they are always the first and last or the second and second to last tokens.

ylemkimon replied:
@kevinbarabash Sorry, my mistake. `end` can be outside the range of tokens, when it's delimited. In that case, `end` is the last delimiter.

kevinbarabash commented on src/MacroExpander.js:
```diff
@@ -122,42 +148,89 @@ export default class MacroExpander implements MacroContextInterface {
+        return {tokens, start, end: tok};
+    }
+
     /**
      * Consume the specified number of arguments from the token stream,
      * and return the resulting array of arguments.
      */
```
Can you update this comment to include that delimiters are consumed if specified?  Also, are there any constraints on the size of `delimiters` based on the value of `numArgs`?  If so, it might be good to add a check to make sure it meets that constraint.

kevinbarabash commented on src/MacroExpander.js:
```diff
@@ -202,8 +271,8 @@ export default class MacroExpander implements MacroContextInterface {
                 "need to increase maxExpand setting");
         }
         let tokens = expansion.tokens;
+        const args = this.consumeArgs(expansion.numArgs, expansion.delimiters);
```
The reason this was a moved up is so that we consume delimiters even if there are no args, is that right?  What's an example of when this could happen?

ylemkimon replied:
@kevinbarabash Any non-alphabetic characters following the control sequence is considered to be delimiters. For instance, `\def\abc|2c{123}` has delimiter `|2c`.

kevinbarabash commented on src/Parser.js:
```diff
@@ -480,31 +473,24 @@ export default class Parser {
-                (i === 0 && !isOptional && this.mode === "math");
+
+            if ((funcData.primitive && argType == null) ||
+                // \sqrt expands into primitive if optional argument doesn't exist
+                (funcData.type === "sqrt" && i === 1 && optArgs[0] == null)) {
```
Is `sqrt` the only macro that does this?

ylemkimon replied:
@kevinbarabash Yes, for functions we currently have and probably most LaTeX macros.

kevinbarabash commented on src/Parser.js:
```diff
@@ -480,31 +473,24 @@ export default class Parser {
-                }
-                throw new ParseError(
-                    `Expected group after '${func}'`, this.fetch());
+                argType, isOptional, baseGreediness);
+            if (isOptional) {
+                optArgs.push(arg);
+            } else if (arg != null) {
+                args.push(arg);
+            } else { // should be unreachable
+                throw new ParseError("Null argument, please report this as a bug");
             }
-            (isOptional ? optArgs : args).push(arg);
         }
```
Thanks for reworking this.  Not having to do a `continue` and getting rid of the ternary is much cleaner.

kevinbarabash commented on src/Parser.js:
```diff
@@ -518,64 +504,50 @@ export default class Parser {
     ): ?AnyParseNode {
         switch (type) {
             case "color":
-                if (consumeSpaces) {
-                    this.consumeSpaces();
-                }
```
Where did all of these calls to `consumeSpaces()` go?

ylemkimon replied:
They have been moved to match TeX behavior: https://github.com/KaTeX/KaTeX/blob/35470fd7db11202e01d3097bbc57fe2a1eb67acd/src/MacroExpander.js#L115-L120 https://github.com/KaTeX/KaTeX/blob/35470fd7db11202e01d3097bbc57fe2a1eb67acd/src/MacroExpander.js#L155-L170

kevinbarabash commented on src/Parser.js:
```diff
@@ -518,64 +504,50 @@ export default class Parser {
             case undefined:
-                return this.parseGroup(name, optional, greediness,
-                    undefined, undefined, consumeSpaces);
+                return this.parseArgumentGroup(optional);
```
Having more specialized functions makes the code easier to understand.

kevinbarabash commented on src/MacroExpander.js:
```diff
@@ -108,6 +108,32 @@ export default class MacroExpander implements MacroContextInterface {
         this.stack.push(...tokens);
     }
 
+    /**
+     * Find an macro argument without expanding tokens and append the array of
+     * tokens to the token stack. Uses Token as a container for the result.
+     */
```
If the argument is a group will the contents of the return token have to be reparsed?

ylemkimon replied:
@kevinbarabash They're not parsed here, but only fetched (buffered).

kevinbarabash commented on src/Parser.js:
```diff
@@ -598,49 +570,20 @@ export default class Parser {
-        this.mode = outerMode;
-        return firstToken.range(lastToken, str);
+        this.consume(); // consume the end of the argument
+        argToken.text = str;
+        return argToken;
```
If we replace the `.text` won't the `argToken.loc` be incorrect?

ylemkimon replied:
@kevinbarabash Normally, they should match since the end of the argument is denoted using `EOF`. Otherwise, it's an error and will be caught by the `EOF` mismatch.

ylemkimon replied:
@kevinbarabash Their lengths may not match, due to whitespace normalization, etc.

kevinbarabash commented on src/Parser.js:
```diff
@@ -705,7 +645,9 @@ export default class Parser {
         let res;
         let isBlank = false;
-        if (!optional && this.fetch().text !== "{") {
+        // don't expand before parseStringGroup
+        this.gullet.consumeSpaces();
+        if (!optional && this.gullet.future().text !== "{") {
```
Is the comment referring to the `if` block?

ylemkimon replied:
@kevinbarabash Yes, but also to `this.gullet.consumeSpaces();`, on why it's used instead of `this.consumeSpaces();`.

kevinbarabash commented on src/Parser.js:
```diff
@@ -678,7 +618,7 @@ export default class Parser {
     parseColorGroup(optional: boolean): ?ParseNode<"color-token"> {
         const res = this.parseStringGroup("color", optional);
-        if (!res) {
+        if (res == null) {
```
This doesn't change the behaviour does it?

ylemkimon replied:
@kevinbarabash It doesn't.

kevinbarabash commented on src/Parser.js:
```diff
@@ -764,52 +706,60 @@ export default class Parser {
-     * at the current position, and ends right before a higher explicit
+     * Parses an argument.
+     */
+    parseArgumentGroup(optional: boolean, mode?: Mode): ?AnyParseNode {
```
Could the result be typed as `?ParseNode<"ordgroup">` or does that cause other issues with the flow types?

ylemkimon replied:
@kevinbarabash Yes, I'll update the PR.

kevinbarabash commented on src/Parser.js:
```diff
@@ -764,52 +706,60 @@ export default class Parser {
-     * If `optional` is false or absent, this parses an ordinary group,
-     * which is either a single nucleus (like "x") or an expression
-     * in braces (like "{x+y}") or an implicit group, a group that starts
-     * at the current position, and ends right before a higher explicit
+     * Parses an argument.
+     */
```
Are arguments the only groups where we specify a mode?  If so, it might be helpful to add a comment saying such.

kevinbarabash commented on src/Parser.js:
```diff
@@ -764,52 +706,60 @@ export default class Parser {
+        }
+
+        this.gullet.beginGroup();
+        const expression = this.parseExpression(false, "EOF");
+        this.expect("EOF"); // expect the end of the argument
```
I wonder if we could use a different token to mark the end of arguments.  Overloading `EOF` seems like it could be a bit confusing.  Maybe something for a future PR given that there's already so much going on in this one.

ylemkimon replied:
@kevinbarabash I agree. This is my major concern, too. This will probably cause a lot of issues that are hard to debug, but I can't seem to find a simple better way.

kevinbarabash commented on src/defineFunction.js:
```diff
@@ -223,6 +228,10 @@ export function defineFunctionBuilders<NODETYPE: NodeType>({
     });
 }
 
+export const normalizeArgument = function(arg: AnyParseNode): AnyParseNode {
+    return arg.type === "ordgroup" && arg.body.length === 1 ? arg.body[0] : arg;
+};
```
Is it possible that `arg` could be a series of nested `ordgroup`s?  If so, would we want to unwrap each of them?

ylemkimon replied:
@kevinbarabash I'm not sure whether it's possible, but that's not caused by changes in this PR.

kevinbarabash commented on src/environments/array.js:
```diff
@@ -61,9 +61,10 @@ function parseArray(
-    // Parse body of array with \\ temporarily mapped to \cr
     parser.gullet.beginGroup();
-    parser.gullet.macros.set("\\\\", "\\cr");
+    // \cr is equivalent to \\ without the optional size argument (see below)
+    // TODO: provide helpful error when \cr is used outside the array environment
```
Can you raise a ticket for this `TODO`?

kevinbarabash commented on src/environments/array.js:
```diff
@@ -125,9 +126,18 @@ function parseArray(
+        } else if (next === "\\\\") {
+            parser.consume();
+            let size;
+            // \def\Let@{\let\\\math@cr}
+            // \def\math@cr{...\math@cr@}
+            // \def\math@cr@{\new@ifnextchar[\math@cr@@{\math@cr@@[\z@]}}
+            // \def\math@cr@@[#1]{...\math@cr@@@...}
+            // \def\math@cr@@@{\cr}
```
Does this collapse multiple `\cr`s in a row or something like that?

ylemkimon replied:
@kevinbarabash No, they are just how `\\` is defined in LaTeX and amsmath.

kevinbarabash replied:
I don't think I'll ever get used to reading .dtx files. 😓

kevinbarabash commented on src/functions/def.js:
```diff
@@ -88,41 +88,65 @@ defineFunction({
+            } else if (tok.text === "EOF") {
+                throw new ParseError("Expected a macro definition");
+            } else {
+                delimiters[numArgs].push(tok.text);
```
Can `tok.text` be any string or should we be checking it against a list of allowed delimiters?

ylemkimon replied:
@kevinbarabash AFAIK, every valid TeX token can be used as delimiters.

kevinbarabash commented on src/functions/def.js:
```diff
@@ -88,41 +88,65 @@ defineFunction({
+                }
+                numArgs++;
+                delimiters.push([]);
+            } else if (tok.text === "EOF") {
+                throw new ParseError("Expected a macro definition");
```
How would this case arise?

ylemkimon replied:
@kevinbarabash If the end is reached without the macro body, e.g., `\def\abc`.

kevinbarabash replied:
This seems really obvious now. 😅

kevinbarabash commented on src/macros.js:
```diff
@@ -240,7 +253,7 @@ defineMacro("\\char", function(context) {
 // TODO: Optional arguments: \newcommand{\macro}[args][default]{definition}
 const newcommand = (context, existsOK: boolean, nonexistsOK: boolean) => {
-    let arg = context.consumeArgs(1)[0];
+    let arg = context.consumeArg().tokens;
```
Returning an object with named properties is much nicer.  Thank you.

kevinbarabash commented on test/errors-spec.js:
```diff
@@ -166,7 +167,7 @@ describe("Parser.expect calls:", function() {
         it("complains about missing { for color", function() {
             expect`\textcolor#ffffff{text}`.toFailWithParseError(
-                   "Expected '{', got '#' at position 11:" +
+                   "Invalid color: '#' at position 11:" +
```
Nice!

kevinbarabash commented on test/errors-spec.js:
```diff
@@ -94,11 +94,12 @@ describe("Parser:", function() {
         it("complains about missing argument at end of input", function() {
             expect`2\sqrt`.toFailWithParseError(
-                   "Expected group after '\\sqrt' at end of input: 2\\sqrt");
+                   "Expected group as argument to '\\sqrt'" +
+                   " at end of input: 2\\sqrt");
```
I like the new wording.

kevinbarabash commented on test/errors-spec.js:
```diff
@@ -176,23 +177,23 @@ describe("Parser.expect calls:", function() {
             expect`\textcolor{#ffffff{text}`.toFailWithParseError(
-                   "Invalid color: '#ffffff{text' at position 12:" +
-                   " \\textcolor{#̲f̲f̲f̲f̲f̲f̲{̲t̲e̲x̲t̲}");
+                   "Unexpected end of input in a macro argument," +
+                   " expected '}' at end of input: …r{#ffffff{text}");
```
Removing the position information may make it hard to track down errors in some cases.  The underlining was also helpful because it made it really easy to see the range of the issue.  Looking at some of the test cases below, we still have both of these features in those test cases.  Is it hard to maintain those features in these test cases because how macro expansion has changed?

ylemkimon replied:
@kevinbarabash In case of macros, arguments should be scanned (fetched) first, so scanning errors are detected before parsing errors, hence changes in error position.

kevinbarabash replied:
I see, this is just how we've been reporting parse for some time now.

kevinbarabash commented on test/errors-spec.js:
```diff
@@ -293,22 +296,22 @@ describe("Lexer:", function() {
             expect`\textcolor{1a2}{foo}`.toFailWithParseError(
                    "Invalid color: '1a2'" +
-                   " at position 12: \\textcolor{1̲a̲2̲}{foo}");
+                   " at position 11: \\textcolor{̲1̲a̲2̲}̲{foo}");
```
Why is there an extra space inside of `\textcolor{}` in the error message?

ylemkimon replied:
@kevinbarabash It is probably combining underline not shown properly.

kevinbarabash replied:
I don't see an extra space anymore. 🤷‍♂️

kevinbarabash commented on test/katex-spec.js:
```diff
@@ -2566,23 +2581,6 @@ describe("A smash builder", function() {
     });
 });
 
-describe("A document fragment", function() {
-    it("should have paddings applied inside an extensible arrow", function() {
-        const markup = katex.renderToString("\\tiny\\xrightarrow\\textcolor{red}{x}");
-        expect(markup).toContain("x-arrow-pad");
-    });
-
-    it("should have paddings applied inside an enclose", function() {
-        const markup = katex.renderToString(r`\fbox\textcolor{red}{x}`);
-        expect(markup).toContain("boxpad");
-    });
-
-    it("should have paddings applied inside a square root", function() {
-        const markup = katex.renderToString(r`\sqrt\textcolor{red}{x}`);
-        expect(markup).toContain("padding-left");
-    });
-});
```
Why were these test cases removed?

ylemkimon replied:
@kevinbarabash They're test cases added in #1707, which relied on edge-case and previous (wrong) behaviors.

kevinbarabash commented on test/katex-spec.js:
```diff
@@ -2990,12 +2983,17 @@ describe("A macro expander", function() {
         }}));
     });
 
+    it("should allow properly nested group for macro argument", function() {
+        expect`\foo{e^{x_{12}+3}}`.toParseLike("(e^{x_{12}+3})",
+            new Settings({macros: {"\\foo": "(#1)"}}));
+    });
```
Nice!

kevinbarabash commented on test/katex-spec.js:
```diff
@@ -3015,9 +3013,8 @@ describe("A macro expander", function() {
         expect`\noexpand\frac xy`.toParseLike`xy`;
-        // TODO(ylem): #2085
         // \def is not expandable, so is not affected by \noexpand
-        // expect`\noexpand\def\foo{xy}\foo`.toParseLike`xy`;
+        expect`\noexpand\def\foo{xy}\foo`.toParseLike`xy`;
```
Yay!

ylemkimon commented on docs/migration.md:
```diff
@@ -0,0 +1,24 @@
+It also no longer accepts replacement text not enclosed in braces. For example,
+`\def\foo1` no longer works and should be changed to `\def\foo{1}`.
+
+## `\newline` and `\cr`
+`\newline` and `\cr` no longer takes an optional size argument. To specify vertical
+spacing, `\\` should be used.
```
@kevinbarabash Partly, yes. Those changed space handling and they are updated to match their definitions.

ronkok replied:
This is a late comment, sorry. But there is a breaking change in that `\raisebox` previously accepted unbraced dimensions and now it does not. That might be worth a note in the migration guide.

ylemkimon replied:
@ronkok Thank you for finding it! That was an unexpected change, but might be fixed with #2139. I'll add a note to the migration guide.

ronkok replied:
I don't think there is anything to "fix".  LaTeX `\raisebox` does not accept unbraced dimensions, so the new behavior is not a bug. It's just needs a note in the migration guide.

ylemkimon commented on src/MacroExpander.js:
```diff
@@ -122,42 +148,89 @@ export default class MacroExpander implements MacroContextInterface {
+        if (!isDelimited) {
+            // Ignore spaces between arguments.  As the TeXbook says:
+            // "After you have said ‘\def\row#1#2{...}’, you are allowed to
+            //  put spaces between the arguments (e.g., ‘\row x n’), because
+            //  TeX doesn’t use single spaces as undelimited arguments."
```
@kevinbarabash They are intended, but if you think they are unnecessary, I'll remove them.

ylemkimon commented on src/functions/cr.js:
```diff
@@ -5,43 +5,29 @@ import defineFunction from "../defineFunction";
-// indicating whether we're in the \cr case, and newLine indicating whether
-// to break the line in the \newline case.
-
+// \DeclareRobustCommand\\{...\@xnewline}
```
@kevinbarabash They are omitted commands for clarity.

ylemkimon commented on src/functions/def.js:
```diff
@@ -88,41 +88,65 @@ defineFunction({
+        while (parser.gullet.future().text !== "{") {
+            tok = parser.gullet.popToken();
+            if (tok.text === "#") {
+                // If the very last character of the <parameter text> is #, so that
+                // this # is immediately followed by {, TeX will behave as if the {
+                // had been inserted at the right end of both the parameter text
+                // and the replacement text.
```
@kevinbarabash https://github.com/KaTeX/KaTeX/blob/35470fd7db11202e01d3097bbc57fe2a1eb67acd/test/katex-spec.js#L3195

kevinbarabash reviewed (approved):
Thanks for answering all my questions and for all of your hard work on this PR.

ylemkimon commented:
@kevinbarabash Thank you for the review!
